### PR TITLE
Implemented see more upcoming/past ride pages + minor homepage fixes

### PR DESCRIPTION
**Pull request by helenxuyang**
### Summary 

This pull request is the first step towards implementing the See More pages for upcoming and past rides. 

See More pages:
- [x] wrote functions for generating cards for repeated rides
- [x] added navigation from homepage to the See More pages

General:
- [x] combined ride cards into one customizable widget for consistency
- [x] fixed RidesProvider to have variables

Homepage:
- [x] fixed shadow on bottom bar of homepage
- [x] fixed padding on sidescrolling upcoming rides page so that it still appears padded but cuts off on the sides when scrolling
- [x] fixed spacing on empty upcoming rides widget
 
### Test Plan 
Upcoming ride cards:

Ride history cards:

See more upcoming rides:

See more ride history:

Unit tests for recurring ride generation are in FutureRideGeneratorTest in the test folder.

### Notes 

- Didn't add lazy loading list because I wasn't sure if it's necessary; will look out for any performance issues in the future and reconsider
- Will add ride editing in a future PR which handles the case when a generated ride card without an ID is edited/deleted; this will need to create a new ride and update the parent ride's edits field, so I added a few parameters for passing around parent IDs but realized editing hasn't been implemented

### CODE REVIEW

File: lib/Home.dart [hunk 1]
```diff
@@ -45,268 +43,245 @@ class Home extends StatelessWidget {
       );
     }
 
-    return Scaffold(
```
Comment (tyang98) [inline lib/Home.dart:48]: Thanks for removing all the unnecessary tabs on the navigation drawer

File: lib/Ride_History.dart [hunk 2]
```diff
@@ -22,82 +20,40 @@ class _RideHistoryState extends State<RideHistory> {
...
-          } else if (snapshot.hasError) {
-            return Text("${snapshot.error}");
-          }
-          return Center(child: CircularProgressIndicator());
-        });
+    List<Ride> pastRides = ridesProvider.pastRides;
+    if (pastRides.length == 0) {
+      return _emptyRideHist();
+    } else {
+      return _mainHist(pastRides);
```
Comment (tyang98) [inline lib/Ride_History.dart:55]: Is there a limit to how many past rides in Ride History we display on the home page?

File: lib/Ride.dart [hunk 3]
```diff
@@ -260,3 +270,224 @@ class Ride {
...
+        Duration rideDuration = originalRide.endTime.difference(originalRide.startTime);
+        List<Ride> deletedInstances = originalRide.edits.map((rideID) => originalRidesByID[rideID]).where((ride) => ride.deleted).toList();
+        List<int> days = originalRide.recurringDays;
+
+        // find first occurrence
+        int daysUntilFirstOccurrence = days.map((day) => daysUntilWeekday(originalRide.startTime, day)).reduce(min);
+        DateTime rideStart = originalRide.startTime.add(Duration(days: daysUntilFirstOccurrence));
+        int dayIndex = days.indexOf(rideStart.weekday);
+
+        // generate instances of recurring rides
```
Comment (tyang98) [inline lib/Ride.dart:418]: The comments throughout all this code really helped explain your implementation :)

File: test/FutureRideGeneratorTest.dart [hunk 4]
```diff
@@ -0,0 +1,290 @@
+// This is a basic Flutter widget test.
```
Comment (tyang98) [inline test/FutureRideGeneratorTest.dart:1]: I'm a bit confused about how I can run these unit tests. Although I know your implementation generates recurring rides properly, would like to know how check these tests.

**Review by tyang98 — CHANGES_REQUESTED** [review]
Overall, your implementation is really sound. There are just some minor changes and questions I have. Great work!

**Review by tyang98 — COMMENTED** [review]
Overall, your implementation is really sound. There are just some minor changes and questions I have. Great work!

**Review by tyang98 — APPROVED** [review]

File: lib/Ride.dart [hunk 5]
```diff
@@ -260,3 +270,224 @@ class Ride {
...
 }
+
+T getOrNull<T>(Map<String, dynamic> map, String key, {T parse(dynamic s)}) {
+  var x = map.containsKey(key) ? map[key] : null;
+  if (x == null) return null;
+  if (parse == null) return x;
+  return parse(x);
+}
+
+class RideCard extends StatelessWidget {
```
[the remark this answers is not in the document]
Comment (helenxuyang) [reply lib/Ride.dart:281]: Fixed this in commit e0ccb4a, thanks for catching that!

File: test/FutureRideGeneratorTest.dart (hunk 4 above)
Comment (helenxuyang) [reply test/FutureRideGeneratorTest.dart:1]: I believe you run **flutter test test/FutureRideGeneratorTest.dart** in the terminal, let me know if that works!

File: lib/Ride_History.dart (hunk 2 above)
Comment (helenxuyang) [reply lib/Ride_History.dart]: Since Aiden's gone we're just gonna pick 5 since that sounds reasonable--just added this, thanks for the catch!

**Pull request merged**